### PR TITLE
Validate array uniqueness

### PR DESCRIPTION
**Pull request by joaoGabriel55**
From discussion: #2316

TODO
- [x] Implement array of primitives uniqueness
- [x]  Implement array of complex objects uniqueness

Thanks for the help: @ramos-ph and @fwoelffel

### CODE REVIEW

**Comment by aarontravass** [comment]
> What suggestions do you have for array of complex objects uniqueness? I have one that would be make the objects elements from array flatten to simplifies the comparison between.

Class-validator seems to be doing a simple comparison

https://github.com/typestack/class-validator/blob/262a3c6369b9a9fc81846a39b62b6a2a21bd8fab/src/decorator/array/ArrayUnique.ts#L7-L20

**Comment by joaoGabriel55** [comment]
@colinhacks could you take a look, when you have a free time?

**Comment by aarontravass** [comment]
@joaoGabriel55 As discussed here, https://github.com/colinhacks/zod/discussions/2316#discussioncomment-6979061, would it be possible to add an options which would display duplicate items?

**Comment by joaoGabriel55** [comment]
> @joaoGabriel55 As discussed here, [#2316 (reply in thread)](https://github.com/colinhacks/zod/discussions/2316#discussioncomment-6979061), would it be possible to add an options which would display duplicate items?

yeah!

File: src/types.ts [hunk 1]
```diff
@@ -2079,6 +2104,22 @@ export class ZodArray<
     return this._def.type;
   }
 
+  _arrayUnique(array: unknown[], identifier?: ArrayUniqueIdentifier) {
```
Comment (aarontravass) [inline src/types.ts:2107]: Can you preferably take an option as input, basically asking if the user wants to display unique elements. Finding duplicates itself is an expensive operation especially if objects are involved. So it is better that we don't display unique elements by default.

**Comment by fwoelffel** [comment]
@colinhacks any thoughts about this?

File: src/types.ts [hunk 2]
```diff
@@ -1975,6 +1975,11 @@ export interface ZodArrayDef<T extends ZodTypeAny = ZodTypeAny>
   exactLength: { value: number; message?: string } | null;
   minLength: { value: number; message?: string } | null;
   maxLength: { value: number; message?: string } | null;
+  uniqueness: {
+    identifier?: ArrayUniqueIdentifier;
```
Comment (fwoelffel) [inline src/types.ts:1979]:
```suggestion
    identifier?: ArrayUniqueIdentifier<z.output<T>>;
```

File: src/types.ts [hunk 3]
```diff
@@ -1975,6 +1975,11 @@ export interface ZodArrayDef<T extends ZodTypeAny = ZodTypeAny>
...
+    message?: string | ((duplicateElements: any) => string);
```
Comment (fwoelffel) [inline src/types.ts:1980]:
```suggestion
    message?: string | ((duplicateElements: z.output<T>[]) => string);
```

File: src/types.ts [hunk 4]
```diff
@@ -2104,6 +2146,23 @@ export class ZodArray<
     return this.min(1, message) as any;
   }
 
+  unique(params?: {
+    identifier?: ArrayUniqueIdentifier;
```
Comment (fwoelffel) [inline src/types.ts:2150]:
```suggestion
    identifier?: ArrayUniqueIdentifier<z.output<T>>;
```

File: src/types.ts [hunk 5]
```diff
@@ -2104,6 +2146,23 @@ export class ZodArray<
...
+    message?: errorUtil.ErrMessage | ((duplicateElements: any) => string);
```
Comment (fwoelffel) [inline src/types.ts:2151]:
```suggestion
    message?: errorUtil.ErrMessage | ((duplicateElements: z.output<T>[]) => string);
```

File: src/types.ts [hunk 6]
```diff
@@ -2104,6 +2146,23 @@ export class ZodArray<
...
+    showDuplicates?: boolean;
+  }): this {
+    return new ZodArray({
+      ...this._def,
+      uniqueness: {
+        ...params,
+        message:
+          typeof params?.message === "function"
+            ? params?.message
```
Comment (fwoelffel) [inline src/types.ts:2160]:
```suggestion
            ? params.message
```

**Review by fwoelffel — CHANGES_REQUESTED** [review]

File: deno/lib/types.ts [hunk 7]
```diff
@@ -1975,6 +1975,11 @@ export interface ZodArrayDef<T extends ZodTypeAny = ZodTypeAny>
   exactLength: { value: number; message?: string } | null;
   minLength: { value: number; message?: string } | null;
   maxLength: { value: number; message?: string } | null;
+  uniqueness: {
+    identifier?: ArrayUniqueIdentifier;
+    message?: string | ((duplicateElements: any) => string);
```
[the remark this answers is not in the document]
Comment (joaoGabriel55) [reply deno/lib/types.ts]: I tried this, but some types of errors like this:
```ts
Type '{ [k in keyof T]: ZodOptional<T[k]>; }' does not satisfy the constraint 'ZodRawShape'.
  Type 'ZodOptional<T[k]>' is not assignable to type 'ZodTypeAny'.
    The types returned by 'refine(...).refine(...).optional()' are incompatible between these types.
      Type 'ZodOptional<ZodEffects<ZodEffects<ZodOptional<T[k]>, any, T[k]["_input"] | undefined>, any, T[k]["_input"] | undefined>>' is not assignable to type 'ZodOptional<ZodEffects<ZodEffects<ZodType<any, any, any>, any, any>, any, any>>'.
        Type 'ZodEffects<ZodType<any, any, any>, any, any>' is not assignable to type 'ZodEffects<ZodOptional<T[k]>, any, T[k]["_input"] | undefined>'.
```
```ts
Type 'Options[number]' does not satisfy the constraint 'ZodType<any, any, any>'.ts(2344)
```

File: deno/lib/types.ts [hunk 8]
```diff
@@ -1986,7 +1986,7 @@ export interface ZodArrayDef<T extends ZodTypeAny = ZodTypeAny>
   maxLength: { value: number; message?: string } | null;
   uniqueness: {
     identifier?: ArrayUniqueIdentifier;
-    message?: string | ((duplicateElements: Array<unknown>) => string);
+    message?: string | ArrayMessageFunction;
```
Comment (fwoelffel) [inline deno/lib/types.ts:1989]: Why don't you use `T` here?
```suggestion
    message?: string | ArrayMessageFunction<T>;
```

File: deno/lib/types.ts [hunk 9]
```diff
@@ -2155,19 +2158,11 @@ export class ZodArray<
...
-      message?:
-        | errorUtil.ErrMessage
-        | ((duplicateElements: Array<output<T>>) => string);
-      showDuplicates?: boolean;
-    } = {}
-  ): this {
+  unique(params: ZodArrayDef<T>["uniqueness"] = {}): this {
     const message =
-      typeof params.message === "function"
+      typeof params?.message === "function"
```
Comment (fwoelffel) [inline deno/lib/types.ts:2163]:
```suggestion
      typeof params.message === "function"
```

File: src/types.ts [hunk 10]
```diff
@@ -1986,7 +1986,7 @@ export interface ZodArrayDef<T extends ZodTypeAny = ZodTypeAny>
   maxLength: { value: number; message?: string } | null;
   uniqueness: {
     identifier?: ArrayUniqueIdentifier;
-    message?: string | ((duplicateElements: Array<unknown>) => string);
+    message?: string | ArrayMessageFunction;
```
Comment (fwoelffel) [inline src/types.ts:1989]: Same question here, is there a specific reason for not using `T`?
```suggestion
    message?: string | ArrayMessageFunction<T>;
```

File: src/types.ts [hunk 11]
```diff
@@ -2155,19 +2158,11 @@ export class ZodArray<
...
-      message?:
-        | errorUtil.ErrMessage
-        | ((duplicateElements: Array<output<T>>) => string);
-      showDuplicates?: boolean;
-    } = {}
-  ): this {
+  unique(params: ZodArrayDef<T>["uniqueness"] = {}): this {
     const message =
-      typeof params.message === "function"
+      typeof params?.message === "function"
```
Comment (fwoelffel) [inline src/types.ts:2163]:
```suggestion
      typeof params.message === "function"
```

File: src/types.ts [hunk 12]
```diff
@@ -2155,19 +2158,11 @@ export class ZodArray<
...
         ? params.message
-        : errorUtil.toString(params.message);
+        : errorUtil.toString(params?.message);
```
Comment (fwoelffel) [inline src/types.ts:2165]:
```suggestion
        : errorUtil.toString(params.message);
```

File: deno/lib/types.ts [hunk 13]
```diff
@@ -2155,19 +2158,11 @@ export class ZodArray<
...
         ? params.message
-        : errorUtil.toString(params.message);
+        : errorUtil.toString(params?.message);
```
Comment (fwoelffel) [inline deno/lib/types.ts:2165]:
```suggestion
        : errorUtil.toString(params.message);
```

**Review by fwoelffel — CHANGES_REQUESTED** [review]

File: deno/lib/types.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (fwoelffel) [reply deno/lib/types.ts]: I'll try to have a look at this during the weekend ;)

File: src/types.ts (hunk 12 above)
Comment (joaoGabriel55) [reply src/types.ts:2165]: I need to put this `?` because the params can be not passed (`undefined`) on the `.unique` method.

Comment (fwoelffel) [reply src/types.ts:2165]: My bad

**Review by fwoelffel — APPROVED** [review]

**Comment by colinhacks** [comment]
Thanks! I like this a lot, especially `identifier`. This will land in Zod 4, but it'll be a while. I'm merging this as a starting point, but it'll undergo a lot of change before then.

**Comment by MentalGear** [comment]
Cool, but I do not get why the most idiomatic approach, z.array(z.string()).unique(), is not supported ?
@joaoGabriel55

**Comment by fwoelffel** [comment]
> Cool, but I do not get why the most idiomatic approach, z.array(z.string()).unique(), is not supported ? @joaoGabriel55

Maybe I'm mistaken, but `z.array(z.string()).unique()` is supported 🤔

**Comment by sethfitz** [comment]
This functionality disappeared somewhere before the v4 release. The merge commit (0e4583d) doesn't exist in the `v4` branch, so I think it got dropped.

**Comment by Abdalrhman-Almarakeby** [comment]
@colinhacks Is there a reason this was dropped? It seems like many people have requested this feature.

**Pull request merged**